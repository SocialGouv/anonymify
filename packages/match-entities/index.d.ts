export type MatchSearchResult = {
  type: MatchEntity;
  score: number;
};

export type MatchSearchResults = MatchSearchResult[];

export type MatchEntity =
  | "address"
  | "city"
  | "date_fr"
  | "date"
  | "datetime"
  | "email"
  | "firstname"
  | "float"
  | "fullname"
  | "geo"
  | "integer"
  | "ip"
  | "json"
  | "lastname"
  | "phone"
  | "postcode"
  | "sex"
  | "siren"
  | "siret"
  | "text"
  | "title"
  | "url";

type CorpusMatcher = {
  url?: string;
  items?: string[];
};

type RegexpsMatcher = {
  regexps: string[];
};

export type Matcher = CorpusMatcher | RegexpsMatcher;

export interface AnonymifyConfig {
  matchers: Partial<Record<MatchEntity>, Matcher>;
}
