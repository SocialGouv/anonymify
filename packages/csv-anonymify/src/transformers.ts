import faker from "faker/locale/fr";
import { format } from "date-fns";

import type { MatchEntity } from "@socialgouv/match-entities";

const leftPad = (s: string | number, length: number = 10) => {
  let padded = ("" + s).slice(0, length);
  while (padded.length < length) {
    padded = "0" + padded;
  }
  return padded;
};

// todo: handle config.items formats
export const transformers = {
  address: (options: any) => faker.address.streetAddress(options),
  city: (options: any) => faker.address.city(options),
  date_fr: (options: any) => format(faker.date.past(), "MM/dd/yyyy"),
  date: (options: any) => format(faker.date.past(), "yyyy-MM-dd"),
  datetime: (options: any) => faker.date.past().toISOString(),
  email: (options: any) => faker.internet.email(options),
  firstname: (options: any) => faker.name.firstName(options),
  float: (options: any) => "" + faker.datatype.float(options),
  fullname: (options: any) => faker.name.findName(options),
  geo: (options: any) =>
    faker.fake("{{address.latitude}},{{address.longitude}}"),
  integer: (options: any) => "" + faker.datatype.number(options),
  ip: (options: any) => faker.internet.ip(),
  json: (options: any) => JSON.stringify({ random: faker.lorem.word() }),
  lastname: (options: any) => faker.name.lastName(options),
  phone: (options: any) => faker.phone.phoneNumber(options),
  postcode: (options: any) =>
    leftPad(faker.datatype.number({ min: 10000, max: 96000 }), 5),
  sex: (options: any) => faker.random.arrayElement(["Homme", "Femme"]),
  siren: (options: any) =>
    leftPad(faker.datatype.number({ min: 100000000, max: 999999999 }), 9),
  siret: (options: any) =>
    leftPad(
      faker.datatype.number({ min: 10000000000000, max: 99999999999999 }),
      14
    ),
  text: (options: any) => faker.lorem.words(options),
  title: (options: any) => faker.random.arrayElement(["M.", "Mme."]),
  url: (options: any) => faker.internet.url(),
} as Record<MatchEntity, any>;
