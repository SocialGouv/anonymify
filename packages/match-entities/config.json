{
  "matchers": {
    "firstname": {
      "corpus": "prenoms"
    },
    "lastname": {
      "corpus": "noms"
    },
    "city": {
      "corpus": "villes"
    },
    "geo": {
      "regexps": ["^\\d\\d?\\.\\d+,\\d\\d?\\.\\d+$"]
    },
    "email": {
      "regexps": ["^[^@/'\";]+@[^@]+\\.[^@]+$"]
    },
    "postcode": {
      "regexps": ["^\\d{5}$", "^2a$", "^2b$"]
    },
    "siren": {
      "regexps": ["^\\d{9}$"]
    },
    "siret": {
      "regexps": ["^\\d{14}$"]
    },
    "phone": {
      "regexps": ["^(\\+\\d+\\s?)?0?\\d[:/.\\s]?(\\d{2}[:/.\\s]?){4}$"]
    },
    "ip": {
      "regexps": ["^\\d\\d?\\d?\\.\\d\\d?\\d?\\.\\d\\d?\\d?\\.\\d\\d?\\d?"]
    },
    "json": {
      "regexps": ["^{\\\".*\\\"\\s*}"]
    },
    "address": {
      "regexps": [
        "^(\\d+\\s*(bis|ter)?[\\s,]+)?(rond[-\\s]point|chemin|route|impasse|place|avenue|av.?|rue|boulevard|bd.?)\\s.*$"
      ]
    },
    "date_fr": {
      "regexps": ["^\\d\\d?/\\d\\d?/\\d\\d(\\d\\d)?$"]
    },
    "date": {
      "regexps": ["^\\d{4}[/-]\\d\\d?[/-]\\d\\d?$"]
    },
    "datetime": {
      "regexps": ["^\\d{4}[/-]\\d\\d?[/-]\\d\\d?T[\\d:.+]+Z?$"]
    },
    "url": {
      "regexps": ["^https?://.*$"]
    },
    "float": {
      "regexps": ["^\\d+[,.]\\d+$"]
    },
    "integer": {
      "regexps": ["^\\d+$"]
    },
    "title": {
      "items": [
        "M",
        "M.",
        "Mr",
        "Mr.",
        "Monsieur",
        "Madame",
        "Mme",
        "Me",
        "Mme.",
        "Mlle."
      ]
    },
    "sex": {
      "items": ["Homme", "H", "Femme", "F"]
    }
  }
}
