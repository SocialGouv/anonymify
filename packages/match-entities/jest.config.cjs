/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  preset: "ts-jest",
  testTimeout: 20000,
  snapshotFormat: {
    printBasicPrototype: false,
  },
};
