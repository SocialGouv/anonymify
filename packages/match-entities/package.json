{
  "name": "@socialgouv/match-entities",
  "decription": "Return entity probability for a given text",
  "version": "1.1.12",
  "license": "Apache-2.0",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist/**"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsup src/index.ts --format esm,cjs --dts --dts-resolve",
    "dev": "tsup src/index.ts --format esm,cjs --watch --dts --dts-resolve",
    "lint": "TIMING=1 eslint src --fix",
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "test": "NODE_OPTIONS=--experimental-vm-modules npx jest"
  },
  "dependencies": {
    "minisearch": "^3.2.0",
    "ngraminator": "^2.0.6"
  },
  "devDependencies": {
    "@tsconfig/recommended": "^1.0.1",
    "@types/flexsearch": "^0.7.2",
    "@types/jest": "^27.0.3",
    "jest": "^27.4.5",
    "ts-jest": "^27.1.2",
    "ts-node": "^10.4.0",
    "tsconfig": "^1.1.12",
    "tsup": "^5.11.9",
    "typescript": "^4.5.4"
  }
}
