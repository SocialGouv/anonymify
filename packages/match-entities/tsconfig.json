{
  "extends": "tsconfig/base",
  "include": ["./src/**/*.ts", "./src/**/*.tsx"],
  "exclude": ["dist", "build", "node_modules"],
  "compilerOptions": {
    "resolveJsonModule": true
  }
}
