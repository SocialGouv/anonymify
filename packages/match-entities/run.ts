import { match } from "./src/index";

(async () => {
  match("juju").then(console.log).catch(console.log);
})();
