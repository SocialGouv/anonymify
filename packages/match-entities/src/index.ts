import { search as fuzzySearch } from "./fuzzy";
import { match as regexMatch } from "./regexps";
//@ts-expect-error
import ngraminator from "ngraminator";

import config from "../config.json";

type SearchResult = {
  type: string;
  score: number;
};

type SearchResults = SearchResult[];

type CorpusMatcher = {
  url?: string;
  items?: string[];
};

type RegexpsMatcher = {
  regexps: string[];
};

type Matcher = CorpusMatcher | RegexpsMatcher;

interface Config {
  matchers: {
    [key: string]: Matcher;
  };
}

const cleanStr = (str: string) =>
  str &&
  str
    .trim()
    .toLowerCase()
    .replace(/[^\w\d]/, "");

const cleanArr = (arr: string[]) => arr.map((v) => cleanStr(v));

const removeStops = (arr: string[]): string[] => {
  const stops = ["le", "de", "la", "du", "d"];
  return arr.map((s) => s.toLowerCase()).filter((s) => stops.indexOf(s) === -1);
};

export const match = async (needle: string): Promise<SearchResults> => {
  if (!needle.trim().length) return [];
  const matches = Object.keys(config.matchers)
    .map((key) => {
      //@ts-expect-error
      const matcher = config.matchers[key] as Matcher;
      if ("regexps" in matcher && regexMatch(needle, matcher.regexps)) {
        return { type: key, score: 100 };
      } else if (
        //@ts-expect-error
        matcher.items &&
        //@ts-expect-error
        cleanArr(matcher.items).indexOf(cleanStr(needle)) > -1
      ) {
        return { type: key, score: 100 };
      }
    })
    .filter(Boolean);

  if (needle.length > 30) {
    matches.push({ type: "text", score: 100 });
  }

  if (!matches.length && needle.length > 3) {
    matches.push(...(await fuzzySearch(needle)));
  }

  if (!matches.length && needle.indexOf(" ") > -1) {
    // search nom + prenom
    const ngrams = ngraminator(needle.split(" "), [1, 2, 3, 4])
      .map((item: string[]) => item.join(" "))
      .filter((s: string) => s.length > 3);

    const ngramResults = (
      await Promise.all(ngrams.map((ngram: string) => fuzzySearch(ngram)))
    )
      .slice(0, 10)
      .flatMap((result: any) => result && result.length && result.slice(0, 5))
      .filter((result: any) => result.score > 0.2) // keep most relevant matches
      .map((result: any) => result.type)
      .filter(Boolean);

    if (
      ngramResults.indexOf("nom") > -1 &&
      ngramResults.indexOf("prenom") > -1
    ) {
      matches.push({ type: "fullname", score: 100 });
    }
  }

  if (!matches.length && !Number(needle)) {
    matches.push({ type: "text", score: 100 });
  }

  return matches as SearchResults;
};
