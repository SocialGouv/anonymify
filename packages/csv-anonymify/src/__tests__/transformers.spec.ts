import faker from "faker/locale/fr";

import { MatchEntity } from "@socialgouv/match-entities";
import { transformers } from "../transformers";

faker.seed(1234);

jest.useFakeTimers().setSystemTime(new Date("2020-01-01").getTime());

const tests = [
  "address",
  "city",
  "date_fr",
  "date",
  "datetime",
  "email",
  "firstname",
  "float",
  "fullname",
  "geo",
  "integer",
  "ip",
  "json",
  "lastname",
  "phone",
  "postcode",
  "sex",
  "siren",
  "siret",
  "text",
  "title",
  "url",
] as MatchEntity[];

tests.forEach((t) => {
  test(`should anonymise ${t}`, () => {
    expect(transformers[t]()).toMatchSnapshot();
  });
});
