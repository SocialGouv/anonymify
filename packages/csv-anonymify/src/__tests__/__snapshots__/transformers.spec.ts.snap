// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should anonymise address 1`] = `"46846 Hincmar Saint-Honor√©"`;

exports[`should anonymise city 1`] = `"Clichy"`;

exports[`should anonymise date 1`] = `"2019-09-22"`;

exports[`should anonymise date_fr 1`] = `"11/07/2019"`;

exports[`should anonymise datetime 1`] = `"2019-10-20T12:58:31.582Z"`;

exports[`should anonymise email 1`] = `"Justin87@hotmail.fr"`;

exports[`should anonymise firstname 1`] = `"Arnaude"`;

exports[`should anonymise float 1`] = `"35781.37"`;

exports[`should anonymise fullname 1`] = `"Mme Ella Da silva"`;

exports[`should anonymise geo 1`] = `"38.2864,108.9503"`;

exports[`should anonymise integer 1`] = `"37025"`;

exports[`should anonymise ip 1`] = `"25.143.14.128"`;

exports[`should anonymise json 1`] = `"{\\"random\\":\\"dignissimos\\"}"`;

exports[`should anonymise lastname 1`] = `"Dubois"`;

exports[`should anonymise phone 1`] = `"0172823768"`;

exports[`should anonymise postcode 1`] = `"16482"`;

exports[`should anonymise sex 1`] = `"Femme"`;

exports[`should anonymise siren 1`] = `"431941605"`;

exports[`should anonymise siret 1`] = `"20569904488511"`;

exports[`should anonymise text 1`] = `"recusandae qui et"`;

exports[`should anonymise title 1`] = `"M."`;

exports[`should anonymise url 1`] = `"http://amandin.net"`;
